### PR TITLE
[FIX] core: maximum recursion because of `active` fields.

### PR DESCRIPTION
**Pull request by ryv-odoo**
## [FIX] core: maximum recursion because of `active` fields.

In specific situation, unlink can lead to raise a `RecursionError`:
- The model `A` has a many2one `b_id` field toward a model `B`. This field is set with `ondelete='cascade'`.
- The model `A` has one **store** related field **no-sudo** named `a_related` (`related='b_id.b_other_field`).
- With `ir.rule` on model `A` with a domain containing `a_related`

You have one record B `b_1` with 20 records A linked to it (`a_1, ..., a_20`). When you try to unlink `b_1`:

### Stack:
```
  File "...", line 543, in ...
    b_1.unlink()
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 3594, in unlink
    self.env.flush_all()
```
=> At this point, `a_1, ..., a_20` have already been deleted from the database because of the 'cascade' deletion. But the ORM doesn't have any information about this, and `a_related` (for `a_1, ..., a_20`) are flagged to be recomputed (because it depends on `b_id.b_other_field`)
```
  File "/home/odoo/Documents/dev/odoo/odoo/api.py", line 732, in flush_all
    self._recompute_all()
  File "/home/odoo/Documents/dev/odoo/odoo/api.py", line 728, in _recompute_all
    self[field.model_name]._recompute_field(field)
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 6165, in _recompute_field
    field.recompute(records)
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 1348, in recompute
    self.compute_value(record)
```
=> `self.compute_value(recs)` raised a `MissingError` before recalling `compute_value` with only the first `record` (but others are still in the prefetch)
```
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 1368, in compute_value
    records._compute_field_value(self)
```
=> `a_related` of `record` is removed from to_compute, but only the first record, not the rest of the records present in the prefetch set.
```
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 4194, in _compute_field_value
    fields.determine(field.compute, self)
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 100, in determine
    return needle(records, *args)
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 689, in _compute_related
    values = [first(value[name]) for value in values]
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 689, in <listcomp>
    values = [first(value[name]) for value in values]
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 5860, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 2772, in __get__
    return super().__get__(records, owner)
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 1186, in __get__
    recs._fetch_field(self)
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 3162, in _fetch_field
    self._read(fnames)
```
=> `_read` tries to read the first record + others from the prefetch set.
```
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 3215, in _read
    self.with_context(active_test=False)._flush_search([], order='id')
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 4607, in _flush_search
    self.env[model_name].flush_model(field_names)
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 5560, in flush_model
    self._recompute_model(fnames)
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 6134, in _recompute_model
    self._recompute_field(field)
```
=> This is where the recursion starts, record compute will move forward one by one. But sadly, the stack grows very fast, and with only a few (already deleted) records to recompute, the issue will be generated.
```
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 6165, in _recompute_field
    field.recompute(records)
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 1348, in recompute
    self.compute_value(record)
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 1368, in compute_value
    records._compute_field_value(self)
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 4194, in _compute_field_value
    fields.determine(field.compute, self)
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 100, in determine
    return needle(records, *args)
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 689, in _compute_related
    values = [first(value[name]) for value in values]
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 689, in <listcomp>
    values = [first(value[name]) for value in values]
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 5860, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 2772, in __get__
    return super().__get__(records, owner)
  File "/home/odoo/Documents/dev/odoo/odoo/fields.py", line 1186, in __get__
    recs._fetch_field(self)
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 3162, in _fetch_field
    self._read(fnames)
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 3215, in _read
    self.with_context(active_test=False)._flush_search([], order='id')
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 4607, in _flush_search
    self.env[model_name].flush_model(field_names)
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 5560, in flush_model
    self._recompute_model(fnames)
  File "/home/odoo/Documents/dev/odoo/odoo/models.py", line 6134, in _recompute_model
    self._recompute_field(field)
```
Fix:
Move the `MissingError` logic of `_recompute_field` inside the `recompute` directly.

## [IMP] core: avoid extra invalidate of compute no-store field.

'tocompute' in the 'Transaction' contains store field on records to be recomputed. No-store compute fields are directly invalidated from the cache when a dependency changes (see 'BaseModel.modified').

In fact, '_recompute_field' was actually doing too much for nothing. Also, it may invalidate caches of compute no-store fields for no reason (e.g., if they are searchable). Remove the part for field compute no-store field. And prevent '_recompute_field' callers from calling it with no-store fields.

opw-3268088

### CODE REVIEW

**Comment by ryv-odoo** [comment]
Thanks @xmo-odoo 

> Not sure it has any relation, but FWIW I encountered a not-entirely-dissimilar issue with stored recursive computed fields, though I mostly worked around it by having `Field.__get__` catch the recursion error & fall back: [8f6f724](https://github.com/odoo/odoo/commit/8f6f724174eaf4fe261576b0657c3875b773519f)

I am pretty sure that your commit (in 16.0) also fixes my problem (maybe not as efficiently, but is it necessary for this kind of corner-case?)

Anyway, it should be great to cover both issues with tests (we can break this kind of thing with a small change in the ORM). Then, I will at least include those tests in this PR.

File: odoo/fields.py [hunk 1]
```diff
@@ -1331,19 +1332,31 @@ def recompute(self, records):
         if not to_compute_ids:
             return
 
+        def missing_guard(func, batch_records):
```
Comment (rco-odoo) [inline odoo/fields.py:1335]: Maybe this one could be a field method, like `field.compute_except_missing(records)`, which calls `field.compute` internally.  Notice that `missing_guard` can only be called with `field.compute_value` or something equivalent to that.

File: odoo/fields.py [hunk 2]
```diff
@@ -1331,19 +1332,31 @@ def recompute(self, records):
...
+                    batch_records.env.remove_to_compute(f, batch_records - existing)
+
         if self.recursive:
-            for record in records:
-                if record.id in to_compute_ids:
-                    self.compute_value(record)
+            def recursive_compute(batch_records):
+                for record in batch_records:
+                    if record.id in to_compute_ids:
+                        self.compute_value(record)
```
Comment (rco-odoo) [inline odoo/fields.py:1348]: I would apply `missing_guard` on each record instead.
```suggestion
            for record in batch_records:
                if record.id in to_compute_ids:
                    missing_guard(self.compute_value, record)
```

**Review by rco-odoo — COMMENTED** [review]
Looks good to me.

File: odoo/fields.py (hunk 2 above)
Comment (ryv-odoo) [reply odoo/fields.py:1348]: It means that we will make one `exists` by record (instead of batch it) :crying_cat_face:

**Review by rco-odoo — APPROVED** [review]
Good job, Rémy!

**Review by rco-odoo — APPROVED** [review]
@robodoo rebase-ff r+

**Pull request closed without merging**